odoo/models.py

```diff
@@ -4952,7 +4974,16 @@ def with_context(self, *args, **kwargs):
             The returned recordset has the same prefetch object as ``self``.
         """
         if args and 'allowed_company_ids' not in args[0] and 'allowed_company_ids' in self._context:
-            args[0]['allowed_company_ids'] = self._context.get('allowed_company_ids') 
+            # Force the companies information to be kept
+            # when context is overridden without allowed_company_ids information.
+            args[0]['allowed_company_ids'] = self._context.get('allowed_company_ids')
+        if args and 'force_company' in args[0]:
+            _logger.warning("Usage of force_company context key has no impact anymore. Use with_company(company) instead.")
+        if args and 'company' in args[0]:
+            _logger.warning(
+                "Use of protected company context key. It may not behave"
+                "as expected and may trigger unneeded recomputations."
```

Comment:
Message: "Context key 'company' is not recommended, because of its special meaning in @depends_context."